pkg/cluster/cluster.go

```diff
@@ -171,36 +196,63 @@ func New(config *rest.Config, opts ...Option) (Cluster, error) {
 	}
 
 	// Create the cache for the cached read client and registering informers
-	cache, err := options.NewCache(config, cache.Options{
-		HTTPClient:  options.HTTPClient,
-		Scheme:      options.Scheme,
-		Mapper:      mapper,
-		ResyncEvery: options.SyncPeriod,
-		Namespaces:  []string{options.Namespace},
-	})
+	cacheOpts := options.Cache
+	{
+		if cacheOpts.Scheme == nil {
+			cacheOpts.Scheme = options.Scheme
+		}
+		if cacheOpts.Mapper == nil {
+			cacheOpts.Mapper = mapper
+		}
+		if cacheOpts.HTTPClient == nil {
+			cacheOpts.HTTPClient = options.HTTPClient
+		}
+		if cacheOpts.ResyncEvery == nil {
+			cacheOpts.ResyncEvery = options.SyncPeriod
+		}
+		if len(cacheOpts.Namespaces) == 0 && options.Namespace != "" {
+			cacheOpts.Namespaces = []string{options.Namespace}
+		}
+	}
+	cache, err := options.NewCache(config, cacheOpts)
 	if err != nil {
 		return nil, err
 	}
 
-	writeObj, err := options.NewClient(config, client.Options{
-		HTTPClient: options.HTTPClient,
-		Scheme:     options.Scheme,
-		Mapper:     mapper,
-		Cache: &client.CacheOptions{
-			Reader:     cache,
-			DisableFor: options.ClientDisableCacheFor,
-		},
-	})
+	// Create the client, and default its options.
+	clientOpts := options.Client
+	{
+		if clientOpts.Scheme == nil {
+			clientOpts.Scheme = options.Scheme
+		}
+		if clientOpts.Mapper == nil {
+			clientOpts.Mapper = mapper
+		}
+		if clientOpts.HTTPClient == nil {
+			clientOpts.HTTPClient = options.HTTPClient
+		}
+		if clientOpts.Cache == nil {
+			clientOpts.Cache.Reader = cache
+			clientOpts.Cache.DisableFor = options.ClientDisableCacheFor
+			clientOpts.Cache.Unstructured = false
+		}
+		if !clientOpts.DryRun && options.DryRunClient {
```

Comment:
maybe make `clientOpts.DryRun` a pointer to be able to differentiate between unset and set here?

Reply:
Good idea, done!